cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +144,125 @@ def __str__(self) -> str:
         return 'CCZ**{}'.format(self._exponent)
 
 
+class ThreeQubitDiagonalGate(gate_features.ThreeQubitGate):
+    """A gate given by a diagonal 8x8 matrix."""
+
+    def __init__(self,
+                 a: float = 0.0,
+                 b: float = 0.0,
+                 c: float = 0.0,
+                 d: float = 0.0,
+                 e: float = 0.0,
+                 f: float = 0.0,
+                 g: float = 0.0,
+                 h: float = 0.0) -> None:
+        self._diag_angles = []  # type: List[float]
+        self._diag_angles.append(a)
+        self._diag_angles.append(b)
+        self._diag_angles.append(c)
+        self._diag_angles.append(d)
+        self._diag_angles.append(e)
+        self._diag_angles.append(f)
+        self._diag_angles.append(g)
+        self._diag_angles.append(h)
+        self._matrix = np.diag(
+            [np.exp(1j * angle) for angle in self._diag_angles])
+
+    def _unitary_(self) -> np.ndarray:
+        return self._matrix
+
+    def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
+                              ) -> protocols.CircuitDiagramInfo:
+        return protocols.CircuitDiagramInfo(('diag', '#2', '#3'))
+
+    def __pow__(self, exponent: Any) -> 'ThreeQubitDiagonalGate':
+        if not isinstance(exponent, (int, float)):
```

Comment:
This will refuse to work on symbolic exponents. (There's probably a lot of places in the code that do this... it used to be a limitation. But it's time to change it now that the limit is gone.)

Allow `sympy.Basic` as one of the allowed types. Use `protocols.mul` instead of `*`, because `*` doesn't do the right thing for expressions like `1.0 * sympy.Symbol('a')`, unfortunately.

Reply:
I removed the check all together (like FSimGate).